src/particle_filter.cpp

```diff
@@ -1,33 +1,95 @@
+    // unisgned interters
+    int min, max;
+    config.value("min_particles", min);
+    config.value("max_particles", max);
+    min_samples_ = min;
+    max_samples_ = max;
+
+    kld_err_ = 0.01;
+    kld_z_ = 0.99;
+    config.value("kld_err", kld_err_, tue::config::OPTIONAL);
+    config.value("kld_z", kld_z_, tue::config::OPTIONAL);
+
+    alpha_slow_ = 0;
+    alpha_fast_ = 0;
+    config.value("recovery_alpha_slow", alpha_slow_, tue::config::OPTIONAL);
+    config.value("recovery_alpha_fast", alpha_fast_, tue::config::OPTIONAL);
+
+    std::array<double, 3> cell_size = {0.5, 0.5, 10*M_PI/180};
+    config.value("cell_size_x", cell_size[0], tue::config::OPTIONAL);
+    config.value("cell_size_y", cell_size[1], tue::config::OPTIONAL);
+    config.value("cell_size_theta", cell_size[2], tue::config::OPTIONAL);
+
+    samples_[0].reserve(max_samples_);
+    samples_[1].reserve(max_samples_);
+
+    cluster_cache_.reserve(max_samples_);
+
+    limit_cache_.clear();
+    limit_cache_.resize(max_samples_, 0);
+
+    kd_tree_.reset(new KDTree(max_samples_, cell_size));
+
+    ROS_INFO_STREAM("[ED Localization] min_samples: " << min_samples_ << ", max_samples: " << max_samples_);
+    ROS_INFO_STREAM("[ED Localization] kld_err: " << kld_err_ << ", kld_z: " << kld_z_);
+    ROS_INFO_STREAM("[ED Localization] recovery_alpha_slow: " << alpha_slow_ << ", recovery_alpha_fast: " << alpha_fast_);
+    ROS_INFO_STREAM("[ED Localization] cell_size_x: " << cell_size[0] << ", cell_size_y: " << cell_size[1] << ", cell_size_theta: " << cell_size[2]);
+}
+
+// ----------------------------------------------------------------------------------------------------
+
 ParticleFilter::~ParticleFilter()
 {
 }
 
 // ----------------------------------------------------------------------------------------------------
 
-void ParticleFilter::initUniform(const geo::Vec2& min, const geo::Vec2& max, double t_step,
-                                 double a_min, double a_max, double a_step)
+void ParticleFilter::initUniform(const geo::Vec2& min, const geo::Vec2& max, double a_min, double a_max)
 {
+    clearCache();
+
     std::vector<Sample>& smpls = samples();
 
     smpls.clear();
-    for(double x = min.x; x < max.x; x += t_step)
-        for(double y = min.y; y < max.y; y += t_step)
-            for(double a = a_min; a < a_max; a += a_step)
+
+    const double range_x = max.x - min.x;
+    const double range_y = max.y - min.y;
+    const double range_yaw = a_max - a_min;
+
+    const double cbrt_samples = ceil(std::cbrt(max_samples_));
+    const double step_x = range_x / cbrt_samples;
+    const double step_y = range_y / cbrt_samples;
+    const double step_yaw = range_yaw / cbrt_samples;
+
+    for(double x = min.x; x < max.x; x += step_x)
+        for(double y = min.y; y < max.y; y += step_y)
+            for(double a = a_min; a < a_max; a += step_yaw)
                 smpls.push_back(Sample(geo::Transform2(x, y, a)));
 
     setUniformWeights();
+
```

Comment:
Would it be possible to insert these into the KD tree directly, without pushing them back into a vector first?

Reply:
The samples are required in the PF itself. The internal structure of the KD tree tracks nodes, not particles. So that will not work.